src/components/tl/mlx5/tl_mlx5_ib.c

```diff
@@ -0,0 +1,467 @@
+                          IBV_QP_ACCESS_FLAGS) != 0) {
+        tl_error(lib, "Failed to modify init qp");
+        goto fail;
+    }
+
+    if (ibv_modify_qp(qp, &qp_attr_to_rtr,
+                      IBV_QP_STATE | IBV_QP_PATH_MTU | IBV_QP_AV |
+                          IBV_QP_MIN_RNR_TIMER) != 0) {
+        tl_error(lib, "Failed to modify init qp");
+        goto fail;
+    }
+
+    *dct_qp = qp;
+    *qpn    = qp->qp_num;
+    return UCC_OK;
+
+fail:
+    if (ibv_destroy_qp(qp)) {
+        tl_error(lib, "Couldn't destroy QP");
+    }
+    return UCC_ERR_NO_MESSAGE;
+}
+
+ucc_status_t ucc_tl_mlx5_init_dci(ucc_tl_mlx5_dci_t *dci, struct ibv_pd *pd,
+                                  struct ibv_context *ctx, struct ibv_cq *cq,
+                                  uint8_t port_num, int tx_depth,
+                                  ucc_base_lib_t *lib)
+{
+    struct ibv_qp_init_attr_ex attr_ex;
+    struct mlx5dv_qp_init_attr attr_dv;
+    struct ibv_qp_attr         qp_attr_to_init;
+    struct ibv_qp_attr         qp_attr_to_rtr;
+    struct ibv_qp_attr         qp_attr_to_rts;
+
+    memset(&attr_ex, 0, sizeof(attr_ex));
+    memset(&attr_dv, 0, sizeof(attr_dv));
+    memset(&qp_attr_to_init, 0, sizeof(qp_attr_to_init));
+    memset(&qp_attr_to_rtr, 0, sizeof(qp_attr_to_rtr));
+    memset(&qp_attr_to_rts, 0, sizeof(qp_attr_to_rts));
+
+    attr_ex.qp_type          = IBV_QPT_DRIVER;
+    attr_ex.send_cq          = cq;
+    attr_ex.recv_cq          = cq;
+    attr_ex.pd               = pd;
+    attr_ex.cap.max_send_wr  = tx_depth;
+    attr_ex.cap.max_send_sge = 1;
+    attr_ex.comp_mask |= IBV_QP_INIT_ATTR_SEND_OPS_FLAGS | IBV_QP_INIT_ATTR_PD;
+    attr_ex.send_ops_flags = IBV_QP_EX_WITH_RDMA_WRITE |
+                             IBV_QP_EX_WITH_RDMA_WRITE_WITH_IMM |
+                             IBV_QP_EX_WITH_ATOMIC_FETCH_AND_ADD;
+    attr_dv.comp_mask |= MLX5DV_QP_INIT_ATTR_MASK_DC |
+                         MLX5DV_QP_INIT_ATTR_MASK_QP_CREATE_FLAGS |
+                         MLX5DV_QP_INIT_ATTR_MASK_SEND_OPS_FLAGS;
+    attr_dv.dc_init_attr.dc_type = MLX5DV_DCTYPE_DCI;
+    attr_dv.create_flags |= MLX5DV_QP_CREATE_DISABLE_SCATTER_TO_CQE;
+
+    attr_dv.send_ops_flags     = MLX5DV_QP_EX_WITH_RAW_WQE;
+    qp_attr_to_init.qp_state   = IBV_QPS_INIT;
+    qp_attr_to_init.pkey_index = 0;
+    qp_attr_to_init.port_num   = port_num;
+    qp_attr_to_init.qp_access_flags =
+        IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ |
+        IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC;
+
+    qp_attr_to_rtr.qp_state          = IBV_QPS_RTR;
+    qp_attr_to_rtr.path_mtu          = IBV_MTU_4096;
+    qp_attr_to_rtr.min_rnr_timer     = 20;
+    qp_attr_to_rtr.ah_attr.port_num  = port_num;
+    qp_attr_to_rtr.ah_attr.is_global = 0;
+
+    qp_attr_to_rts.qp_state      = IBV_QPS_RTS;
+    qp_attr_to_rts.timeout       = 10;
+    qp_attr_to_rts.retry_cnt     = 7;
+    qp_attr_to_rts.rnr_retry     = 7;
```

Comment:
i would suggest to make define for numbers e.g.
```
#define UCC_QP_RNR_RETRY 7
#define UCC_QP_TIMEOUT 10
#define UCC_QP_ACCESS_FLAGS (IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC)
```
they are used in multiple places